File: swift/1099_TwoSumLessThanK.swift
class Solution {
    func twoSumLessThanK(_ A: [Int], _ K: Int) -> Int {
        var sortedNums = A.sorted { $0 < $1 }
        var closestSum = -1
        
        var left = 0, right = sortedNums.count - 1
        while left < right {
            var candidateSum = sortedNums[left] + sortedNums[right]
            if candidateSum > K {
                right -= 1
            } else {
                if candidateSum < K { closestSum = max(candidateSum, closestSum) }
                left += 1
            }
        }
        return closestSum
    }
}

File: java/1099_TwoSumLessThanK.java
class Solution {
    public int twoSumLessThanK(int[] A, int K) {
        int[] counts = new int[1001];
        for (int n : A) {
            counts[n] += 1;
        }
        
        int closeSum = -1;
        int left = 1, right = 1000;
        while (left < right) {
            if (left + right >= K || counts[right] == 0) {
                right--;
            } else {
                if (counts[left] > 0) 
                    closeSum = Math.max(left + right, closeSum);
                left++;
            }
        }
        return closeSum;
    }
}
